joint_trajectory_controller/test/test_trajectory_controller.cpp

```diff
@@ -1578,9 +1600,20 @@ TEST_P(TrajectoryControllerTestParameterized, test_ignore_partial_old_trajectory
   publish(time_from_start, points_old, rclcpp::Time());
   expected_actual.positions = {points_old[0].begin(), points_old[0].end()};
   expected_desired = expected_actual;
-  //  Check that we reached end of points_old[0]trajectory
-  auto end_time =
-    waitAndCompareState(expected_actual, expected_desired, executor, rclcpp::Duration(delay), 0.1);
+  traj_controller_->wait_for_trajectory(executor);
+  auto end_time = updateControllerAsync(
+    rclcpp::Duration(delay), rclcpp::Time(0, 0, RCL_STEADY_TIME),
+    rclcpp::Duration::from_seconds(1.0 / traj_controller_->get_update_rate()));
+
+  if (traj_controller_->has_position_command_interface())
+  {
+    //  Check that we reached end of points_old[0] trajectory
+    auto state_feedback = traj_controller_->get_state_feedback();
+    for (size_t i = 0; i < expected_actual.positions.size(); ++i)
+    {
+      EXPECT_NEAR(expected_actual.positions[i], state_feedback.positions[i], 0.01);
+    }
+  }
```

Comment:
Is there a reason that you don't call waitAndCompareState() here? why is now `traj_controller_->has_position_command_interface()` necessary, but wasn't before?

Reply:
If you look at the implementation of `waitAndCompareState`, you'll see that it is almost equivalent:

```c++
    traj_controller_->wait_for_trajectory(executor);
    auto end_time = updateControllerAsync(controller_wait_time, start_time);

    // get states from class variables
    auto state_feedback = traj_controller_->get_state_feedback();
    auto state_reference = traj_controller_->get_state_reference();

    for (size_t i = 0; i < expected_actual.positions.size(); ++i)
    {
      SCOPED_TRACE("Joint " + std::to_string(i));
      // TODO(anyone): add checking for velocities and accelerations
      if (traj_controller_->has_position_command_interface())
      {
        EXPECT_NEAR(expected_actual.positions[i], state_feedback.positions[i], allowed_delta);
      }
    }

    for (size_t i = 0; i < expected_desired.positions.size(); ++i)
    {
      SCOPED_TRACE("Joint " + std::to_string(i));
      // TODO(anyone): add checking for velocities and accelerations
      if (traj_controller_->has_position_command_interface())
      {
        EXPECT_NEAR(expected_desired.positions[i], state_reference.positions[i], allowed_delta);
      }
    }
```

* I preferred giving a specific update rate to the `updateControllerAsync` function that matches the actual update rate.
* The `waitAndCompareState` function only checks positions, hence I also only checked positions. So, it wasn't necessary before, because it was *inside* `waitAndCompareState`.
* More importantly, `waitAndCompareState` checks for the actual position and the generated position command. However, due to spline interpolation we don't necessarily know the generated command. Assuming `expected_desired = expected_actual;` is simply wrong as long as the robot is in motion (which it is, continuing to the second waypoint). It did work, more or less by accident, since the robot is initialized at `{1.2, 2.1, 3.1}` and moves to `{2.0, 3.0, 4.0}` in the first segment. Hence, the trajectory "arrives" at the setpoint quite slowly. However, in the next segment (which has the same time), it is supposed to go to `{4.0, 5.0, 6.0}`, so it has to move much faster in that segment. Before my change, sampling was basically moved back one cycle, so the desired state was 2.0 for joint1 which matches the desired state and the actual state was something right before that, which was in the range of the 0.1 tolerance (which is probably the reason why it was so large). However, after the changes from this PR, the **actual** state will be at 2.0 at this time and the **desired** will be on the way towards 4.0 already. (In fact, it will be at 2.3599999999999999). If we move sampling back by one control cycle to `rclcpp::Duration(delay) - rclcpp::Duration::from_seconds(0.1)` the desired state matches 2.0 exactly, while the actual state is 1.982, which falls into the 0.1 range. Since the comment to that check says `Check that we reached end of points_old[0]trajectory` I thought it would be better to check exactly that with a small tolerance instead of having to increase the tolerance to actually not fail a check that doesn't only check what we want to know at that point. I hope, that paragraph makes sense...

Reply:
Ignore the question of `has_position_command_interface`, I have overlooked this in the implementation of `waitAndCompareState`.
The rest seems to be reasonable, thanks!